src/capella/execution.rs

```diff
@@ -0,0 +1,294 @@
+    const MAX_TRANSACTIONS_PER_PAYLOAD: usize,
+> {
+    pub parent_hash: Hash32,
+    pub fee_recipient: ExecutionAddress,
+    pub state_root: Bytes32,
+    pub receipts_root: Bytes32,
+    pub logs_bloom: ByteVector<BYTES_PER_LOGS_BLOOM>,
+    pub prev_randao: Bytes32,
+    #[serde(with = "crate::serde::as_string")]
+    pub block_number: u64,
+    #[serde(with = "crate::serde::as_string")]
+    pub gas_limit: u64,
+    #[serde(with = "crate::serde::as_string")]
+    pub gas_used: u64,
+    #[serde(with = "crate::serde::as_string")]
+    pub timestamp: u64,
+    pub extra_data: ByteList<MAX_EXTRA_DATA_BYTES>,
+    pub base_fee_per_gas: U256,
+    pub block_hash: Hash32,
+    pub transactions_root: Root,
+    pub withdrawals_root: Root,
+}
+
+impl<
+        'a,
+        const BYTES_PER_LOGS_BLOOM: usize,
+        const MAX_EXTRA_DATA_BYTES: usize,
+        const MAX_BYTES_PER_TRANSACTION: usize,
+        const MAX_TRANSACTIONS_PER_PAYLOAD: usize,
+        const MAX_WITHDRAWALS_PER_PAYLOAD: usize,
+    >
+    TryFrom<
+        &'a mut ExecutionPayload<
+            BYTES_PER_LOGS_BLOOM,
+            MAX_EXTRA_DATA_BYTES,
+            MAX_BYTES_PER_TRANSACTION,
+            MAX_TRANSACTIONS_PER_PAYLOAD,
+            MAX_WITHDRAWALS_PER_PAYLOAD,
+        >,
+    >
+    for ExecutionPayloadHeader<
+        BYTES_PER_LOGS_BLOOM,
+        MAX_EXTRA_DATA_BYTES,
+        MAX_BYTES_PER_TRANSACTION,
+        MAX_TRANSACTIONS_PER_PAYLOAD,
+    >
+{
+    type Error = Error;
+
+    fn try_from(
+        payload: &'a mut ExecutionPayload<
+            BYTES_PER_LOGS_BLOOM,
+            MAX_EXTRA_DATA_BYTES,
+            MAX_BYTES_PER_TRANSACTION,
+            MAX_TRANSACTIONS_PER_PAYLOAD,
+            MAX_WITHDRAWALS_PER_PAYLOAD,
+        >,
+    ) -> Result<
+        ExecutionPayloadHeader<
+            BYTES_PER_LOGS_BLOOM,
+            MAX_EXTRA_DATA_BYTES,
+            MAX_BYTES_PER_TRANSACTION,
+            MAX_TRANSACTIONS_PER_PAYLOAD,
+        >,
+        Self::Error,
+    > {
+        let transactions_root = payload.transactions.hash_tree_root()?;
+        let withdrawals_root = payload.withdrawals.hash_tree_root()?;
+
+        Ok(ExecutionPayloadHeader {
+            parent_hash: payload.parent_hash.clone(),
+            fee_recipient: payload.fee_recipient.clone(),
+            state_root: payload.state_root.clone(),
+            receipts_root: payload.receipts_root.clone(),
+            logs_bloom: payload.logs_bloom.clone(),
+            prev_randao: payload.prev_randao.clone(),
+            block_number: payload.block_number,
+            gas_limit: payload.gas_limit,
+            gas_used: payload.gas_used,
+            timestamp: payload.timestamp,
+            extra_data: payload.extra_data.clone(),
+            base_fee_per_gas: payload.base_fee_per_gas.clone(),
+            block_hash: payload.block_hash.clone(),
+            transactions_root,
+            withdrawals_root,
+        })
+    }
+}
+
+pub trait ExecutionEngine<
```

Comment:
I haven't thought about how to make the execution bit forward-compatible... I'll come back to this

Reply:
if anything `ExecutionEngine` should just be generic over some type `E` to represent the execution payload but I'll just add an issue to track this